### DISCUSSION TITLE
How should we deal with different workload characteristic?

### DISCUSSION THREAD

**Discussion by sunya-ch** [sustainable-computing-io/kepler-model-server#92] [Ideas]
This discussion will be an extension discussion regarding the comment during the project biweekly meeting (March 07, 2023). (i) Should we separate power model for different workload characteristic (cpu-intensive, memory-intensive, so on)?
If yes, how could we do that?

In my opinion, it depends on how we classify the workload. 
If we classify it by aggregated resource usage, it may not be much different from having a model using resource usage as a feature. (in the other words, there may be no need to separate the power model for different workload characteristic).

However, if we come up with other idea that can classify the workload more particularly than the aggregated resource usage such as usage pattern and observe the different impact to the power consumption, it might be worth separating the model. 

What do you think?

**Comment by rootfs** [comment]
I think the more details we can get from the workload, the more accurate the model can calculate the power consumption.

We currently have CPU cycles/instructions and cache misses. These don't capture the entire nature of the workload. For instance, CPU instructions don't tell the complexity of the operation (and thus the amount of energy), cache misses don't tell if there are memory accesses. 

I think it is time we consider adding two more features: CPU uops, last level cache misses, so that we can at least know the complexity of the CPU instruction complexity (i.e. CPU intensive) and whether there are memory accesses (i.e. memory-intensive).

**Comment by marceloamaral** [comment]
I believe we don't need to develop separate models for distinct workloads. However, it is necessary to train the model with various workloads.

The issue is determining when to retrain the model and how to recognize if the workload differs from those previously utilized for training.

Additionally, @rootfs, it would be advantageous to include additional metrics, not only new hardware counters but also some OS performance counters. There are many of them. This study [here](https://www.microsoft.com/en-us/research/wp-content/uploads/2012/11/iiswc12.pdf) uses only OS metrics and also introduces a feature selection mechanism. We could have something similar in the future, as the work says, different metrics are more or less correlated to the power consumption in different servers.

**Comment by rootfs** [comment]
Sounds good @marceloamaral Let's look into the OS level counters to find a good set of features
